src/gtk/toga_gtk/widgets/table.py

```diff
@@ -23,7 +25,28 @@ def create(self):
         self.native.interface = self.interface
 
     def refresh(self):
-        raise NotImplementedError()
+        self.data.clear()
+
+        for row in self.interface.data.rows:
+            self.data.append(row.data)
 
     def set_on_select(self, handler):
-        pass
+
+        for conn_id in self._connections:
+            # Disconnect all other on_select handlers, so that if you reassign
+            # the on_select, it doesn't trigger the old ones too.
+            self.table.disconnect(conn_id)
+
+        if handler is None:
+            return
+
+        def _handler(widget, *args):
+            selection = widget.get_selection()
+            selection.set_mode(Gtk.SelectionMode.SINGLE)
+            tree_model, tree_iter = selection.get_selected()
```

Comment:
You are missing one edge case. If you select a row and then delete it, `tree_iter` is `None`. This results in a TypeError because  `tree_model.get_path(tree_iter)` does not accept `None` as a argument. So `row` should return the row number or `None` if no row is selected.